doni/driver/worker/blazar.py

```diff
@@ -1,20 +1,270 @@
+            except BlazarAPIError as exc:
+                # TODO what error code does blazar return if the host has a lease already?
+                if exc.code == 404:
+                    host = _search_hosts_for_uuid(hardware.uuid)
+                    if host:
+                        # update stored host_id with match, and retry after defer
+                        result["blazar_host_id"] = host.get("id")
+                    else:
+                        # remove invalid stored host_id and retry after defer
+                        result["blazar_host_id"] = None
+                    return WorkerResult.Defer(result)
+                else:
+                    raise
+        else:
+            # We don't have a cached host_id, try to create a host. If the host exists,
+            # blazar will match the uuid, and the request will fail.
+            try:
+                host = _call_blazar(
+                    context,
+                    f"/os-hosts",
+                    method="post",
+                    json=_blazar_host_requst_body(hardware),
+                    allowed_status_codes=[201],
+                )
+                result["blazar_host_id"] = host.get("id")
+                result["host_created_at"] = host.get("created_at")
+            except BlazarAPIError as exc:
+                if exc.code == 404:
+                    # host isn't in ironic?
+                    return WorkerResult.Defer(result)
+                if exc.code == 409:
+                    host = _search_hosts_for_uuid(hardware.uuid)
+                    if host:
+                        # update stored host_id with match, and retry after defer
+                        result["blazar_host_id"] = host.get("id")
+                    else:
+                        # got conflict despite no matching host,
+                        raise BlazarIsWrongError
+                    return WorkerResult.Defer(result)
+                else:
+                    raise
+
+        # List of all leases from blazar.
+        lease_list_response = _call_blazar(
+            context,
+            "/leases",
+            method="get",
+            allowed_status_codes=[200],
+        )
+        print(f"lease_list_response: {lease_list_response}")
+
+        # Loop over all availability windows for this hw item that Doni has
+        for aw in availability_windows or []:
+
+            aw_dict = _blazar_lease_requst_body(aw)
+            # Check to see if lease name already exists in blazar
+            matching_lease = next(
+                (
+                    lease
+                    for lease in lease_list_response.get("leases")
+                    if lease.get("name") == aw_dict.get("name")
+                ),
+                None,
+            )
+
+            # If the lease exists, then see if it needs to be updated
+            if matching_lease:
+                if not (aw_dict.items() <= matching_lease.items()):
```

Comment:
What is this checking, exactly? I can't reason very well about what the `<=` operator does for two lists of tuples

Reply:
this is checking if the "dict to send in an update" is a subset of "dict of properties of lease".
If it's a subset, then we don't need to call update.
(not sure how to "remove" a key though...)